rfc/0018-certificate-handling.md

```diff
@@ -0,0 +1,259 @@
+
+At the beginning of the reconciliation loop, the code should check the expiration date of the
+certificate used by the Policy Server. If the certificate is about to expire a new one should be
+created. Check the previous section to see which actions have to be done in this circumstance.
+
+The loop should also monitor changes done to the secret that holds the internal CA.
+The reconciliation loop should then make sure that the Policy Server certificate has been signed by
+our CA. If not, a new certificate should be issued (like during a renewal).
+
+> Note: checking if the Policy Server certificate has been issued by our CA might be computationally intense.
+> We have to ensure this is not going to cause a CPU spike. We could resort to do something clever using
+> the revision number of the Secret that holds the CA.
+
+### ClusterAdmissionPolicy
+
+The reconciliation loop should watch also the Secret object that contains the root CA certificate.
+
+During the reconciliation loop, the code must ensure the
+`(Validating|Mutating)WebhookConfiguration` object related with the `ClusterAdmissionPolicy`
+has the `clientConfig.caBundle` properly configured. The field must container the certificate of the internal root CA.
+
+### AdmissionPolicy
+
+The reconciliation loop should watch also the Secret object that contains the root CA certificate.
+
+During the reconciliation loop, the code must ensure the
+`(Validating|Mutating)WebhookConfiguration` object related with the `AdmissionPolicy`
+has the `clientConfig.caBundle` properly configured. The field must container the certificate of the internal root CA.
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+No significant drawback is introduced by this RFC. Our codebase already handles the certificate
+creation for all the Policy Server instances.
+
+On the other hand, we are not yet handling certificate rotation. Hence the implementation of this
+RFC will improve the end user experience and make Kubewarden more robust.
+
+# Alternatives
+
+[alternatives]: #alternatives
+
+Unfortunately there are no alternatives, since delegating certificate management to CertManager
+is not an option. Even if that was an option, we would still have to deal with some aspects
+of certificate management like handling the `(Validating|Mutating)WebhookConfiguration`
+resources.
+
+# Unresolved questions
+
+[unresolved]: #unresolved-questions
+
+The renewal of the root CA is the most disruptive event that can happen. Letting the reconciler loops run and wait
+for the system to converge towards the desired state is not a viable option. During the convergence
+time there could be requests being rejected due to Policy Server instances not being available or be using
+a certificate signed by an unexpected root CA.
+
+We must do some research and find a way to reduce this disruption. However, given our root CA has an expiration
+time of 10 years, this is a problem that should not block the implementation of this RFC.
```

Comment:
I did some research, it's possible to renew a CA without changing its private key. In this way all the certificates issued by it are still going to be valid.
However, this is sometimes considered not a good security practice, it's recommended to create a new private key to: 1) push forward the cracking attempts of an attacker 2) eventually upgrade to a bigger key, better algorithm.

There's [this page](https://docs.aws.amazon.com/privateca/latest/userguide/ca-lifecycle.html) on AWS docs (search for "Reissuing an old CA") that provides this explanation, but I found similar opinions over the internet.

In our context, I think we should not worry too much about this scenario...

However, an easier solution to this problem could be the following one:

* create a new root CA
* create a CA bundle that contains the old CA and the new one
* update all the validation/mutation webook configurations to use the new CA bundle
* generate new certificates for all the policy servers, this time using the new root CA
* rollout the new certificates to the individual policy servers
* wait for the rollout to be complete
* create a new CA bundle that has only the new root CA
* update all the validation/mutation webook configurations to use the new CA bundle

I think this should work. What do you think? Let's discuss over there, I'll then incorporate the final outcome into the RFC

Reply:
I agree. This is a better approach. Let's do that.

Reply:
Done with https://github.com/kubewarden/rfc/pull/24/commits/de8307cf3abad09d0a65a9674b2d301775b62d53